File: API/2441_MasterMind_3/CS/MasterMind3Strategy.cs
using System;
using System.Linq;
using System.Collections.Generic;

using Ecng.Common;
using Ecng.Collections;
using Ecng.Serialization;

using StockSharp.Algo.Indicators;
using StockSharp.Algo.Strategies;
using StockSharp.BusinessEntities;
using StockSharp.Messages;

namespace StockSharp.Samples.Strategies;

/// <summary>
/// Williams %R based extreme reversals.
/// Enters long when all indicators show deep oversold.
/// Enters short when all indicators show strong overbought.
/// </summary>
public class MasterMind3Strategy : Strategy
{
	private readonly StrategyParam<int> _wprPeriod1;
	private readonly StrategyParam<int> _wprPeriod2;
	private readonly StrategyParam<int> _wprPeriod3;
	private readonly StrategyParam<int> _wprPeriod4;
	private readonly StrategyParam<DataType> _candleType;

	/// <summary>
	/// Period for the first Williams %R indicator.
	/// </summary>
	public int WprPeriod1
	{
		get => _wprPeriod1.Value;
		set => _wprPeriod1.Value = value;
	}

	/// <summary>
	/// Period for the second Williams %R indicator.
	/// </summary>
	public int WprPeriod2
	{
		get => _wprPeriod2.Value;
		set => _wprPeriod2.Value = value;
	}

	/// <summary>
	/// Period for the third Williams %R indicator.
	/// </summary>
	public int WprPeriod3
	{
		get => _wprPeriod3.Value;
		set => _wprPeriod3.Value = value;
	}

	/// <summary>
	/// Period for the fourth Williams %R indicator.
	/// </summary>
	public int WprPeriod4
	{
		get => _wprPeriod4.Value;
		set => _wprPeriod4.Value = value;
	}

	/// <summary>
	/// The type of candles to use for calculations.
	/// </summary>
	public DataType CandleType
	{
		get => _candleType.Value;
		set => _candleType.Value = value;
	}

	/// <summary>
	/// Initializes strategy parameters.
	/// </summary>
	public MasterMind3Strategy()
	{
		_wprPeriod1 = Param(nameof(WprPeriod1), 26)
			.SetGreaterThanZero()
			.SetDisplay("WPR Period 1", "Length of the first Williams %R indicator", "WilliamsR")
			
			.SetOptimize(10, 50, 5);

		_wprPeriod2 = Param(nameof(WprPeriod2), 27)
			.SetGreaterThanZero()
			.SetDisplay("WPR Period 2", "Length of the second Williams %R indicator", "WilliamsR")
			
			.SetOptimize(10, 50, 5);

		_wprPeriod3 = Param(nameof(WprPeriod3), 29)
			.SetGreaterThanZero()
			.SetDisplay("WPR Period 3", "Length of the third Williams %R indicator", "WilliamsR")
			
			.SetOptimize(10, 50, 5);

		_wprPeriod4 = Param(nameof(WprPeriod4), 30)
			.SetGreaterThanZero()
			.SetDisplay("WPR Period 4", "Length of the fourth Williams %R indicator", "WilliamsR")
			
			.SetOptimize(10, 50, 5);

		_candleType = Param(nameof(CandleType), TimeSpan.FromMinutes(5).TimeFrame())
			.SetDisplay("Candle Type", "Type of candles for the strategy", "General");
	}

	/// <inheritdoc />
	public override IEnumerable<(Security sec, DataType dt)> GetWorkingSecurities()
	{
		return [(Security, CandleType)];
	}

	/// <inheritdoc />
	protected override void OnStarted2(DateTime time)
	{
		base.OnStarted2(time);

		var wpr1 = new WilliamsR { Length = WprPeriod1 };
		var wpr2 = new WilliamsR { Length = WprPeriod2 };
		var wpr3 = new WilliamsR { Length = WprPeriod3 };
		var wpr4 = new WilliamsR { Length = WprPeriod4 };

		var subscription = SubscribeCandles(CandleType);
		subscription
			.Bind(wpr1, wpr2, wpr3, wpr4, ProcessCandle)
			.Start();

		var area = CreateChartArea();
		if (area != null)
		{
			DrawCandles(area, subscription);
			DrawIndicator(area, wpr1);
			DrawIndicator(area, wpr2);
			DrawIndicator(area, wpr3);
			DrawIndicator(area, wpr4);
			DrawOwnTrades(area);
		}
	}

	private void ProcessCandle(ICandleMessage candle, decimal wpr1, decimal wpr2, decimal wpr3, decimal wpr4)
	{
		if (candle.State != CandleStates.Finished)
			return;

		if (!IsFormedAndOnlineAndAllowTrading())
			return;

		var isBuySignal = wpr1 <= -80m && wpr2 <= -80m && wpr3 <= -80m && wpr4 <= -80m;
		var isSellSignal = wpr1 >= -20m && wpr2 >= -20m && wpr3 >= -20m && wpr4 >= -20m;

		if (isBuySignal && Position <= 0)
		{
			BuyMarket(Volume + Math.Abs(Position));
		}
		else if (isSellSignal && Position >= 0)
		{
			SellMarket(Volume + Math.Abs(Position));
		}
	}
}

File: API/2441_MasterMind_3/PY/master_mind3_strategy.py
import clr

clr.AddReference("StockSharp.Messages")
clr.AddReference("StockSharp.Algo")
clr.AddReference("StockSharp.Algo.Indicators")
clr.AddReference("StockSharp.Algo.Strategies")

from System import TimeSpan
from StockSharp.Messages import DataType, CandleStates, Unit, UnitTypes
from StockSharp.Algo.Indicators import RelativeStrengthIndex
from StockSharp.Algo.Strategies import Strategy


class master_mind3_strategy(Strategy):
    def __init__(self):
        super(master_mind3_strategy, self).__init__()

        self._rsi_period1 = self.Param("RsiPeriod1", 26)
        self._rsi_period2 = self.Param("RsiPeriod2", 27)
        self._rsi_period3 = self.Param("RsiPeriod3", 29)
        self._rsi_period4 = self.Param("RsiPeriod4", 30)
        self._candle_type = self.Param("CandleType", DataType.TimeFrame(TimeSpan.FromMinutes(15)))

        self._was_oversold = False
        self._was_overbought = False

    @property
    def RsiPeriod1(self):
        return self._rsi_period1.Value

    @RsiPeriod1.setter
    def RsiPeriod1(self, value):
        self._rsi_period1.Value = value

    @property
    def RsiPeriod2(self):
        return self._rsi_period2.Value

    @RsiPeriod2.setter
    def RsiPeriod2(self, value):
        self._rsi_period2.Value = value

    @property
    def RsiPeriod3(self):
        return self._rsi_period3.Value

    @RsiPeriod3.setter
    def RsiPeriod3(self, value):
        self._rsi_period3.Value = value

    @property
    def RsiPeriod4(self):
        return self._rsi_period4.Value

    @RsiPeriod4.setter
    def RsiPeriod4(self, value):
        self._rsi_period4.Value = value

    @property
    def CandleType(self):
        return self._candle_type.Value

    @CandleType.setter
    def CandleType(self, value):
        self._candle_type.Value = value

    def OnStarted2(self, time):
        super(master_mind3_strategy, self).OnStarted2(time)

        self._was_oversold = False
        self._was_overbought = False

        rsi1 = RelativeStrengthIndex()
        rsi1.Length = self.RsiPeriod1
        rsi2 = RelativeStrengthIndex()
        rsi2.Length = self.RsiPeriod2
        rsi3 = RelativeStrengthIndex()
        rsi3.Length = self.RsiPeriod3
        rsi4 = RelativeStrengthIndex()
        rsi4.Length = self.RsiPeriod4

        subscription = self.SubscribeCandles(self.CandleType)
        subscription.Bind(rsi1, rsi2, rsi3, rsi4, self.ProcessCandle).Start()

    def ProcessCandle(self, candle, rsi1_val, rsi2_val, rsi3_val, rsi4_val):
        if candle.State != CandleStates.Finished:
            return

        if not self.IsFormedAndOnlineAndAllowTrading():
            return

        r1 = float(rsi1_val)
        r2 = float(rsi2_val)
        r3 = float(rsi3_val)
        r4 = float(rsi4_val)

        is_oversold = r1 <= 35.0 and r2 <= 35.0 and r3 <= 35.0 and r4 <= 35.0
        is_overbought = r1 >= 65.0 and r2 >= 65.0 and r3 >= 65.0 and r4 >= 65.0

        buy_signal = is_oversold and not self._was_oversold
        sell_signal = is_overbought and not self._was_overbought

        if buy_signal and self.Position <= 0:
            self.BuyMarket()
        elif sell_signal and self.Position >= 0:
            self.SellMarket()

        self._was_oversold = is_oversold
        self._was_overbought = is_overbought

    def OnReseted(self):
        super(master_mind3_strategy, self).OnReseted()
        self._was_oversold = False
        self._was_overbought = False

    def CreateClone(self):
        return master_mind3_strategy()
